chain/network-primitives/src/types.rs

```diff
@@ -311,14 +314,75 @@ pub enum NetworkViewClientMessages {
     EpochSyncRequest { epoch_id: EpochId },
     /// A request for headers and proofs during Epoch Sync
     EpochSyncFinalizationRequest { epoch_id: EpochId },
-    /// Get Chain information from Client.
-    GetChainInfo,
     /// Account announcements that needs to be validated before being processed.
     /// They are paired with last epoch id known to this announcement, in order to accept only
     /// newer announcements.
     AnnounceAccount(Vec<(AnnounceAccount, Option<EpochId>)>),
 }
 
+#[derive(Debug)]
+pub struct WithHash<T> {
+    inner: T,
+    hash: u64,
+}
+
+impl<T: Hash> Hash for WithHash<T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_u64(self.hash);
+    }
+}
+
+impl<T: Hash + Default> Default for WithHash<T> {
+    fn default() -> Self {
+        WithHash::new(T::default())
+    }
+}
+
+impl<T> std::cmp::PartialEq for WithHash<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
+    }
+}
+
+impl<T> std::ops::Deref for WithHash<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T: Hash> WithHash<T> {
+    pub fn new(inner: T) -> Self {
+        let mut hasher = std::collections::hash_map::DefaultHasher::new();
+        inner.hash(&mut hasher);
+        Self { inner, hash: hasher.finish() }
+    }
+}
+
+/// Set of account keys with a cached hash.
+/// BTreeMap implements Hash (while HashMap doesn't).
+pub type AccountKeys = WithHash<BTreeMap<(EpochId, AccountId), PublicKey>>;
```

Comment:
The point is to minimize the number of assumptions about the client <-> network communication to make the code less error prone. Here I want make the set_key efficient no matter how efficient is Client in filtering out duplicate calls.

WithHash has been already removed for a different reason (in the previous PR; it will be visible after a force push).

Reply:
Also for a decent bug-resistance I'd like the trigger condition for when to send SetChainInfo the simplest possible (and relatively frequent). Simple condition, means that it will be harder to make a bug in that condition. "at every new chain head" sounds like a good candidate to me. And I strongly recommend sending all the data every time (i.e. also the epoch data, even if it didn't change), since if we send the information only when it actually changes, we rely on the multiple assumptions (for correctness):
1. there is no race condition on the client side when sending the message
2. actix will deliver the messages in the same order that they are send
3. actix won't drop any message on the floor
4. PeerManager will handle the message synchronously
This is a lot of places to make a mistake by accident in the future. And we don't need any of that, if we just send it regularly (for example, at every block, so every ~1s ), since then the information will just eventually stabilize in the periods that it doesn't change.

These are the regular concerns in communication between isolated agents. It would be a different story if there was no actix and if PeerManager exposed an API directly with atomic calls.

To make sending SetChainInfo even more resistant to these kind of problems, we can version the TIER1 account set, by using EpochHeight for example - the difference between EpochId and EpochHeight is that we have a total order on EpochHeight.